src/overlays/actors/ovl_En_Bubble/z_en_bubble.c

```diff
@@ -23,45 +26,393 @@ const ActorInit En_Bubble_InitVars = {
+    sp78.y = this->actor.posRot.pos.y + this->actor.shape.unk_08;
+    sp78.z = this->actor.posRot.pos.z;
+    sp6C = sp78;
+
+    sp6C.x += (sp54.x * 24.0f);
+    sp6C.y += (sp54.y * 24.0f);
+    sp6C.z += (sp54.z * 24.0f);
+    if (func_8003DE84(&globalCtx->colCtx, &sp78, &sp6C, &sp84, &sp94, 1, 1, 1, 0, &sp4C) != 0) {
+        sp60.x = sp94->norm.x * 0.00003051851f;
+        sp60.y = sp94->norm.y * 0.00003051851f;
+        sp60.z = sp94->norm.z * 0.00003051851f;
+        EnBubble_Vec3fNormalizedRelfect(&sp54, &sp60, &sp54);
+        this->bounceDirection = sp54;
+        bounceCount = this->bounceCount;
+        this->bounceCount = ++bounceCount;
+        if (bounceCount > (s16)(Rand_ZeroOne() * 10.0f)) {
+            this->bounceCount = 0;
+        }
+        bounceSpeed = (this->bounceCount == 0) ? 3.6000001f : 3.0f;
+        this->velocityFromBump.x = this->velocityFromBump.y = this->velocityFromBump.z = 0.0f;
+        this->velocityFromBounce.x = (this->bounceDirection.x * bounceSpeed);
+        this->velocityFromBounce.y = (this->bounceDirection.y * bounceSpeed);
+        this->velocityFromBounce.z = (this->bounceDirection.z * bounceSpeed);
+        this->sinkSpeed = 0.0f;
+        Audio_PlayActorSound2(&this->actor, NA_SE_EN_AWA_BOUND);
+        this->graphicRotSpeed = 128.0f;
+        this->graphicEccentricity = 0.48f;
+    } else if (this->actor.bgCheckFlags & 0x20 && sp54.y < 0.0f) {
+        sp60.x = sp60.z = 0.0f;
+        sp60.y = 1.0f;
+        EnBubble_Vec3fNormalizedRelfect(&sp54, &sp60, &sp54);
+        this->bounceDirection = sp54;
+        bounceCount = this->bounceCount;
+        this->bounceCount = ++bounceCount;
+        if (bounceCount > (s16)(Rand_ZeroOne() * 10.0f)) {
+            this->bounceCount = 0;
+        }
+        bounceSpeed = (this->bounceCount == 0) ? 3.6000001f : 3.0f;
+        this->velocityFromBump.x = this->velocityFromBump.y = this->velocityFromBump.z = 0.0f;
+        this->velocityFromBounce.x = (this->bounceDirection.x * bounceSpeed);
+        this->velocityFromBounce.y = (this->bounceDirection.y * bounceSpeed);
+        this->velocityFromBounce.z = (this->bounceDirection.z * bounceSpeed);
+        this->sinkSpeed = 0.0f;
+        Audio_PlayActorSound2(&this->actor, NA_SE_EN_AWA_BOUND);
+        this->graphicRotSpeed = 128.0f;
+        this->graphicEccentricity = 0.48f;
+    }
+    this->actor.velocity.x = this->velocityFromBounce.x + this->velocityFromBump.x;
+    this->actor.velocity.y = this->velocityFromBounce.y + this->velocityFromBump.y + this->sinkSpeed;
+    this->actor.velocity.z = this->velocityFromBounce.z + this->velocityFromBump.z;
+    Math_ApproachF(&this->velocityFromBump.x, 0.0f, 0.3f, 0.1f);
+    Math_ApproachF(&this->velocityFromBump.y, 0.0f, 0.3f, 0.1f);
+    Math_ApproachF(&this->velocityFromBump.z, 0.0f, 0.3f, 0.1f);
+}
+
+u32 func_809CC648(EnBubble* this) {
+    if (((this->colliderSphere.base.acFlags & 0x2) != 0) == false) {
+        return 0;
+    }
```

Comment:
what in the world is this lmfao

Reply:
This is the cause of much pain.

I think maybe the (flags & 2) != 0 could have been a macro, in isolation it seems to gen the same instructions as !!(flags & 2). If that was a macro it becomes slightly more reasonable but still so dumb